### PR TITLE
Refactor!: Simplify PARSENAME <-> SPLIT_PART transpilation

### PR DESCRIPTION
**Pull request by VaggelisD**
Follow up on #4211

### CODE REVIEW

File: sqlglot/dialects/tsql.py [hunk 1]
```diff
@@ -326,24 +326,21 @@ def _parse(args: t.List[exp.Expression]) -> exp.Expression:
...
-    part_num = int(arg_partnum.name)
-    length = 1 if isinstance(arg_this, exp.Null) else part_count + 1  # Reverse index
-    idx = 0 if isinstance(arg_this, exp.Null) else part_num
-    return exp.SplitPart(
-        this=arg_this, delimiter=exp.Literal.string("."), part_num=exp.Literal.number(length - idx)
-    )
+    # PARSENAME(...) will be stored into exp.SplitPart if:
+    # - All args are literals
+    # - The part index (2nd arg) is <= 4 (max valid value, otherwise TSQL returns NULL)
+    if len(args) == 2 and all(isinstance(arg, exp.Literal) for arg in args):
```
Comment (georgesittas) [inline sqlglot/dialects/tsql.py:332]: Can also do `all(arg.is_string for arg in args)` here.

File: sqlglot/dialects/tsql.py [hunk 2]
```diff
@@ -978,27 +975,25 @@ def lateral_op(self, expression: exp.Lateral) -> str:
...
-            arg_this: exp.Expression = expression.args["this"]
-            arg_partnum: exp.Expression = expression.args["part_num"]
-            if isinstance(arg_this, exp.Column) or isinstance(arg_partnum, exp.Column):
+            this = expression.this
+            split_count = len(this.name.split("."))
+            delimiter = expression.args.get("delimiter")
+            part_index = expression.args.get("part_index")
+
+            if (
+                not all(isinstance(arg, exp.Literal) for arg in (this, delimiter, part_index))
```
Comment (georgesittas) [inline sqlglot/dialects/tsql.py:984]: Ditto

**Review by georgesittas — APPROVED** [review]

File: sqlglot/dialects/tsql.py (hunk 1 above)
Comment (VaggelisD) [reply sqlglot/dialects/tsql.py:332]: We don't want them all to be strings, the 2nd arg of `PARSENAME` (and 3rd arg of `SPLIT_PART`) is an integer

Comment (georgesittas) [reply sqlglot/dialects/tsql.py:332]: Ah I see, my bad.

**Pull request merged**